### ISSUE TITLE
[Feature discussion] Add variants through plugins

### ISSUE THREAD

**Issue by adamwathan** [tailwindcss/tailwindcss#496]
Was brainstorming a bit on what it might look like to support adding new variants (hover, focus, etc) through the plugin system.

Here's a bit of code exploring what that API might look like:

```js
function ({ addVariant }) {
  // Simple selector change
  addVariant('even', function (className) {
    return `.even\\:${className}:nth-child(even)`
  })

  // Multi-part selector
  addVariant('group-hover', function (className) {
    return `.group:hover .group-hover\\:${className}`
  })

  // Another multi-part selector
  addVariant('children', function (className) {
    return `.children\\:${className} > *`
  })

  // Change selector and modify properties
  addVariant('important', function (className, properties) {
    return {
      [`.${className}\\!`]: {
        ..._.map(properties, (value) => `${value} !important`)
      },
    }
  })

  // Variant that needs to all be wrapped in an at-rule (note optional third arg)
  addVariant('supports-grid', function (className) {
    return `.supports-grid\\:${className}`
  }, '@supports (display: grid)')
}
```

Anything anyone can think of that would be useful to support that this API couldn't handle?

**Comment by benface** [comment]
Would it be possible to define where the variant rules will appear in the compiled CSS? Maybe at least an option between before and after the default rules?

**Comment by adamwathan** [comment]
> Would it be possible to define where the variant rules will appear in the compiled CSS? Maybe at least an option between before and after the default rules?

So as part of this feature, my plan is to make generated variant order mirror the order in the `modules` section of the config file.

So if you have:

```
modules: {
  // ...
  backgroundColors: ['responsive', 'hover', 'focus'],
  borderColors: ['responsive', 'focus', 'hover'],
}
```

...`focus` variants would be generated before `hover` variants for border colors, but the opposite would be true for background colors. The `responsive` variant is sort of a special case in that those variants are always inserted at the end of your CSS (or wherever you've put the `@tailwind screens` at-rule), but everything else would follow the order in your config for maximum control and flexibility.

**Comment by benface** [comment]
@adamwathan So all variants would be generated after the default/normal rules? Unless I’m misunderstanding, that would be a problem for the `children` variant, since you wouldn’t be able to override a child’s style as I explained in #277. Example:

```
<ul class="list-reset children:block children:rounded-full children:bg-black">
    <li>
        1
    </li>
    <li class="bg-red"> <!-- bg-red will be overridden by children:bg-black -->
        2 (selected)
    </li>
    <li>
        3
    </li>
</ul>
```

**Comment by adamwathan** [comment]
I think we could solve that by adding an optional `normal`/`plain`/`default` variant that you could add to the list to control where the unmodified utilities are generated, and if it's not present we just push it onto the beginning of the list.

**Comment by benface** [comment]
@adamwathan You always come up with the most thoughtful solutions. :)

**Comment by Anachron** [comment]
You have even ... why not have `odd` as well?

**Comment by adamwathan** [comment]
This isn't about adding `even`, it's about adding the ability for people to create their own variants, and I'm using `even` as an example to explain the API.

**Comment by Anachron** [comment]
Sorry very new to tailwindcss and didn't understand the new API at first. Yeah,- totally would be useful!

**Comment by Benecke** [comment]
Boy oh boy, I'm exited! Came here looking for exactly this!

**tldr:** I need different layouts based on app states which are defined by a css class on a parent element (e.g. body element).

**Details:**

I've got a big header on my site for desktop. On tablet and downwards this header will become a minimal version of itself. Mobile nav menu, no searchbar, you name it. No problem as this can be done with tailwind.

The header has two states: **static and fixed position**. So I threw in the some JS for recalculation/offset and when fixed gave the body a class: .has-fixed-header. Now while the header is in that fixed position state, I'd like to reduce it to the minimal version (to save vertical screen space).

I'm currently doing those extensions myself:

```
.hidden-has-fixed-header {
  .has-fixed-header & {
    @apply .hidden;
  }
}

.block-has-fixed-header {
  .has-fixed-header & {
    @apply .block;
  }
}
```

So yes, @adamwathan, **totally upvoting your idea** here!
Is it even possible to use your code right now?

Again, so happy I met tailwind (great name by the way)!
